src/cargo/core/gc.rs

```diff
@@ -0,0 +1,550 @@
+    let factor = match right {
+        "second" | "seconds" => 1,
+        "minute" | "minutes" => 60,
+        "hour" | "hours" => 60 * 60,
+        "day" | "days" => 24 * 60 * 60,
+        "week" | "weeks" => 7 * 24 * 60 * 60,
+        "month" | "months" => 30 * 24 * 60 * 60,
+        _ => return None,
+    };
+    Some(Duration::from_secs(factor * count))
+}
+
+/// Parses a time span string.
+pub fn parse_time_span(span: &str) -> CargoResult<Duration> {
+    maybe_parse_time_span(span).ok_or_else(|| {
+        format_err!(
+            "expected a value of the form \
+             \"N seconds/minutes/days/weeks/months\", got: {span:?}"
+        )
+    })
+}
+
+/// Parses a file size using metric or IEC units.
+pub fn parse_human_size(input: &str) -> CargoResult<u64> {
+    let re = regex::Regex::new(r"(?i)^([0-9]+(\.[0-9])?) ?(b|kb|mb|gb|kib|mib|gib)?$").unwrap();
+    let cap = re.captures(input).ok_or_else(|| {
+        format_err!(
+            "invalid size `{input}`, \
+             expected a number with an optional B, kB, MB, GB, kiB, MiB, or GiB suffix"
+        )
+    })?;
+    let factor = match cap.get(3) {
+        Some(suffix) => match suffix.as_str().to_lowercase().as_str() {
+            "b" => 1.0,
+            "kb" => 1_000.0,
+            "mb" => 1_000_000.0,
+            "gb" => 1_000_000_000.0,
+            "kib" => 1024.0,
+            "mib" => 1024.0 * 1024.0,
+            "gib" => 1024.0 * 1024.0 * 1024.0,
+            s => panic!("suffix `{s}` out of sync with regex"),
+        },
+        None => {
+            return cap[1]
+                .parse()
+                .with_context(|| format!("expected an integer size, got `{}`", &cap[1]))
+        }
+    };
+    let num = cap[1]
+        .parse::<f64>()
+        .with_context(|| format!("expected an integer or float, found `{}`", &cap[1]))?;
+    Ok((num * factor) as u64)
+}
+
+/// Performs automatic garbage collection.
+///
+/// This is called in various places in Cargo where garbage collection should
+/// be performed automatically based on the config settings. The default
+/// behavior is to only clean once a day.
+///
+/// This should only be called in code paths for commands that are already
+/// doing a lot of work. It should only be called *after* crates are
+/// downloaded so that the last-use data is updated first.
+///
+/// It should be cheap to call this multiple times (subsequent calls are
+/// ignored), but try not to abuse that.
+pub fn auto_gc(config: &Config) {
+    if !config.cli_unstable().gc {
+        return;
+    }
+    if !config.network_allowed() {
+        // As a conservative choice, auto-gc is disabled when offline. If the
+        // user is indefinitely offline, we don't want to delete things they
+        // may later depend on.
+        return;
+    }
+
+    if let Err(e) = auto_gc_inner(config) {
+        if global_cache_tracker::is_silent_error(&e)
+            && config.shell().verbosity() != Verbosity::Verbose
```

Comment:
Currently "extra verbose" is a property of `config` and not `shell`.  If we refactored that, this would break.